### PR TITLE
pubsub: no longer block on Ack() calls

### LINKED ISSUE #678: pubsub: no longer block on Ack() calls
As per @rsc's comment, it doesn't make sense to do different things conditionally on whether an Ack call succeeds, so we may as well remove the `error` return and no longer block in these calls.

### PR DESCRIPTION
**Pull request by ijt**
Fixes #678

### CODE REVIEW

File: internal/pubsub/pubsub.go [hunk 1]
```diff
@@ -42,23 +42,15 @@ type Message struct {
 // be sent again to the associated Subscription. This method blocks until
```
Comment (vangent) [inline internal/pubsub/pubsub.go]: Comment is stale.

File: internal/pubsub/pubsub.go [hunk 2]
```diff
@@ -42,23 +42,15 @@ type Message struct {
...
-	}
-	return <-mec.errChan
-}
-
-type msgErrChan struct {
-	msg     *Message
-	errChan chan error
+	go func() {
+		if err := m.sub.ackBatcher.Add(m, size); err != nil {
+			panic(err)
```
Comment (vangent) [inline internal/pubsub/pubsub.go]: `panic` doesn't seem like a good idea in a library... What are the reasons for `Add` to return an `error`? Maybe just

`go m.sub.ackBatcher.Add(m, size)`

?

File: internal/pubsub/pubsub.go [hunk 3]
```diff
@@ -244,19 +241,18 @@ func (s *Subscription) Close() error {
...
-		for _, mec := range mecs {
-			m := mec.msg
+		for _, m := range ms {
 			id := m.ackID
 			ids = append(ids, id)
 		}
 		err := d.SendAcks(ctx, ids)
-		for _, mec := range mecs {
-			mec.errChan <- err
+		if err != nil {
```
Comment (vangent) [inline internal/pubsub/pubsub.go]: Definitely do not want to panic here. 

Maybe `driver.SendAcks` should not return an error anymore, if we're not going to use it for anything....? Or maybe the concrete type should retry...?

File: internal/pubsub/pubsub.go (hunk 2 above)
Comment (ijt) [reply internal/pubsub/pubsub.go]: The reasons are `ErrOversizedItem` or `ErrOverflow`.

File: internal/pubsub/pubsub.go [hunk 4]
```diff
@@ -247,9 +243,12 @@ func NewSubscription(ctx context.Context, d driver.Subscription, opts *Subscript
...
 		}
-		err := d.SendAcks(ctx, ids)
-		if err != nil {
-			panic(err)
+		for {
+			err := d.SendAcks(ctx, ids)
+			if err == nil {
+				break
+			}
+			time.Sleep(time.Second)
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/pubsub/pubsub.go]: Done.

File: internal/pubsub/pubsub.go (hunk 1 above)
Comment (ijt) [reply internal/pubsub/pubsub.go]: Freshened.

File: internal/pubsub/acks_test.go [hunk 5]
```diff
@@ -162,39 +158,16 @@ func TestTooManyAcksForASingleBatchGoIntoMultipleBatches(t *testing.T) {
...
-
-func TestCancelAck(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
 	m := &driver.Message{}
 	ds := &ackingDriverSub{
 		q: []*driver.Message{m},
 		sendAcks: func(_ context.Context, ackIDs []driver.AckID) error {
-			<-ctx.Done()
-			return ctx.Err()
+			c := make(chan struct{})
```
Comment (jba) [inline internal/pubsub/acks_test.go]: You can write this as
```
select {}
```

File: internal/pubsub/pubsub.go [hunk 6]
```diff
@@ -202,17 +188,14 @@ func (s *Subscription) Close() error {
...
-			m := mec.msg
+		for _, m := range ms {
 			id := m.ackID
 			ids = append(ids, id)
 		}
-		err := d.SendAcks(ctx, ids)
-		for _, mec := range mecs {
-			mec.errChan <- err
-		}
+		// TODO(#695): Do something sensible if SendAcks returns an error.
```
Comment (jba) [inline internal/pubsub/pubsub.go]: As Russ suggested, we should probably return an error from Receive. (In a separate PR.)

File: internal/pubsub/acks_test.go (hunk 5 above)
Comment (ijt) [reply internal/pubsub/acks_test.go]: Done.

File: internal/pubsub/pubsub.go (hunk 6 above)
Comment (ijt) [reply internal/pubsub/pubsub.go]: I'm not so sure about that. Anyway, let's move this discussion to #695.

File: internal/pubsub/acks_test.go [hunk 7]
```diff
@@ -75,13 +74,15 @@ func TestAckTriggersDriverSendAcksForOneMessage(t *testing.T) {
...
 func TestMultipleAcksCanGoIntoASingleBatch(t *testing.T) {
 	ctx := context.Background()
+	var wg sync.WaitGroup
 	sentAcks := make(map[driver.AckID]int)
 	ids := []int{1, 2}
 	ds := &ackingDriverSub{
 		q: []*driver.Message{{AckID: ids[0]}, {AckID: ids[1]}},
 		sendAcks: func(_ context.Context, ackIDs []driver.AckID) error {
 			for _, id := range ackIDs {
 				sentAcks[id]++
```
Comment (jba) [inline internal/pubsub/acks_test.go]: We probably shouldn't assume that sendAcks is called single-threaded. That will depend on details of the bundler/batcher. So I'd add a mutex.

Comment (ijt) [reply internal/pubsub/acks_test.go]: Done.

File: internal/pubsub/pubsub.go [hunk 8]
```diff
@@ -39,9 +39,7 @@ type Message struct {
 }
 
 // Ack acknowledges the message, telling the server that it does not need to
-// be sent again to the associated Subscription. This method blocks until
-// the message has been confirmed as acknowledged on the server, or failure
-// occurs.
+// be sent again to the associated Subscription.
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/pubsub/pubsub.go]: I left a comment about this on the issue, let's continue there.

[the remark this answers is not in the document]
Comment (ijt) [reply internal/pubsub/pubsub.go]: Done.

**Comment by ijt** [comment]
Ping, @vangent, can you approve?

**Comment by vangent** [comment]
> Ping, @vangent, can you approve?

Not yet; we are discussing on the issue.

File: internal/pubsub/pubsub.go [hunk 9]
```diff
@@ -202,19 +194,16 @@ func (s *Subscription) Close() error {
...
-		mecs := item.([]msgErrChan)
+		ms := item.([]*Message)
 		var ids []driver.AckID
-		for _, mec := range mecs {
-			m := mec.msg
+		for _, m := range ms {
 			id := m.ackID
 			ids = append(ids, id)
 		}
-		err := d.SendAcks(ctx, ids)
```
Comment (vangent) [inline internal/pubsub/pubsub.go]: It looks like this is no longer calling `SendAcks` at all...?

Comment (ijt) [reply internal/pubsub/pubsub.go]: Huh? There is a call to SendAcks still.

Comment (vangent) [reply internal/pubsub/pubsub.go]: Ah sorry, it is down below, I missed it.

**Pull request merged**